libkbfs/rekey_queue.go

```diff
@@ -36,188 +35,92 @@ import (
+			} else {
+				rkq.config.KBFSOps().RequestRekey(id)
+			}
+			go func(id tlf.ID) {
+				rkq.mu.Lock()
+				defer rkq.mu.Unlock()
+				delete(rkq.pendings, id)
+			}(id)
 		}
-	}
-}
-
-func (rkq *RekeyQueueStandard) ensureRunningLocked() {
-	if rkq.cancel != nil {
-		return
-	}
-	var ctx context.Context
-	ctx, rkq.cancel = context.WithCancel(context.Background())
-	for i := 0; i < numRekeyWorkers; i++ {
-		go rkq.work(ctx)
-	}
-}
-
-func (rkq *RekeyQueueStandard) getRekeyChannelRLocked(id tlf.ID) <-chan error {
-	if ch, ok := rkq.pendings[id]; ok {
-		return ch
-	}
-	return nil
-}
-
-func (rkq *RekeyQueueStandard) getRekeyChannel(id tlf.ID) <-chan error {
-	rkq.lock.RLock()
-	defer rkq.lock.RUnlock()
-	return rkq.getRekeyChannelRLocked(id)
+	}()
 }
 
 // Enqueue implements the RekeyQueue interface for RekeyQueueStandard.
-func (rkq *RekeyQueueStandard) Enqueue(id tlf.ID) <-chan error {
-	rkq.log.Debug("Enqueueing %s for rekey", id)
-
-	if ch := rkq.getRekeyChannel(id); ch != nil {
-		return ch
-	}
-
-	rkq.lock.Lock()
-	defer rkq.lock.Unlock()
-
-	// Now we are locked, check again in case another one slips in.
-	if ch := rkq.getRekeyChannelRLocked(id); ch != nil {
-		return ch
-	}
-
-	rkq.ensureRunningLocked()
-
-	rkq.wg.Add(1)
-
-	ch := make(chan error, 1)
-	rkq.pendings[id] = ch
+func (rkq *RekeyQueueStandard) Enqueue(id tlf.ID) {
+	rkq.mu.Lock()
+	defer rkq.mu.Unlock()
+	rkq.pendings[id] = true
 
 	select {
-	case rkq.queue <- rekeyQueueEntry{id: id, ch: ch}:
+	case rkq.queue <- id:
 	default:
 		// The queue is full; avoid blocking by spawning a goroutine.
 		rkq.log.Debug("Rekey queue is full; enqueuing %s in the background", id)
-		go func() { rkq.queue <- rekeyQueueEntry{id: id, ch: ch} }()
+		go func() { rkq.queue <- id }()
 	}
-
-	return ch
 }
 
 // IsRekeyPending implements the RekeyQueue interface for RekeyQueueStandard.
 func (rkq *RekeyQueueStandard) IsRekeyPending(id tlf.ID) bool {
-	rkq.lock.RLock()
-	defer rkq.lock.RUnlock()
-	_, ok := rkq.pendings[id]
-	return ok
+	rkq.mu.RLock()
+	defer rkq.mu.RUnlock()
+	return rkq.pendings[id]
 }
 
-// Clear implements the RekeyQueue interface for RekeyQueueStandard.
-func (rkq *RekeyQueueStandard) Clear() {
-	rkq.lock.Lock()
-	defer rkq.lock.Unlock()
-	if rkq.cancel == nil {
-		return
-	}
-
+// Shutdown implements the RekeyQueue interface for RekeyQueueStandard.
+func (rkq *RekeyQueueStandard) Shutdown() {
+	rkq.mu.Lock()
+	defer rkq.mu.Unlock()
 	rkq.cancel()
-	for more := true; more; { // drain rkq.queue
-		select {
-		case e := <-rkq.queue:
-			rkq.doneLocked(e, context.Canceled)
-		default:
-			more = false
-		}
-	}
 	rkq.cancel = nil
 }
 
-// Wait implements the RekeyQueue interface for RekeyQueueStandard.
-func (rkq *RekeyQueueStandard) Wait(ctx context.Context) error {
-	return rkq.wg.Wait(ctx)
+// New implements the RekeyQueue interface for RekeyQueueStandard.
+func (rkq *RekeyQueueStandard) New() RekeyQueue {
```

Comment:
It seems confusing to have this instance construct its successor -- why not just have the caller do these two things?